### PR TITLE
Composite Characters Fix

### LINKED ISSUE #1530: Composite japanese characters cause files to be not found
Experiencing a bug related to encoding problems with japanese characters with filenames in Yume2kki:

***
#### Player platform:
macOS 10.13.6

#### Locale:
`ru_RU.UTF-8`

#### Save File and Log
[save-and-log.zip](https://github.com/EasyRPG/Player/files/2651405/save-and-log.zip)

#### Steps to reproduce:
Navigate to track No. 003 D in the PC music player.
***

In particular, the Player claims to not be able to find 夢オルゴールFC.

Interestingly enough, [grepping the music folder](https://pastebin.com/puBHLCUj) has shown that the track does not indeed exist under that name, but exists under the name 夢オルゴールFC instead.

Looks similar right?

Hence I have further copy/pasted the correct and the wrong name into text files and ran xxd against both of them producing:
```
$ xxd correct.txt
00000000: e5a4 a2e3 82aa e383 abe3 82b3 e382 99e3  ................
00000010: 83bc e383 ab46 430a                      .....FC.
```
```
$ xxd wrong.txt
00000000: e5a4 a2e3 82aa e383 abe3 82b4 e383 bce3  ................
00000010: 83ab 4643 0a                             ..FC.
```
which are clearly different.
```
wrong: e5a4a2e382aae383ab e382b4        e383bce383ab46430a
right: e5a4a2e382aae383ab e382b3 e38299 e383bce383ab46430a
```
In particular the difference is in the ゴ character as shown by [grepping the music folder](https://pastebin.com/puBHLCUj).

Potentially related issues are #1383 and #839

### PR DESCRIPTION
**Pull request by SmiVan**
Closes #1530 and probably #1383 and #839, by implementing normalization in place of lower-casing per @Ghabry's patch in #839.

Has to be accepted together with [the related liblcf pull request](https://github.com/EasyRPG/liblcf/pull/292) (EasyRPG/liblcf#292).

Requires more thorough testing, but is confirmed to work against the Yume2kki example in #1530.

***
Also needs to be tested that it doesn't break everything on [non-Mac platforms which use NFC as opposed to NFD](https://gist.github.com/JamesChevalier/8448512).

### CODE REVIEW

**Comment by carstene1ns** [comment]
Just a reminder for later: this might break some existing emscripten games (json cache).

**Comment by carstene1ns** [comment]
ok to test.

**Comment by Ghabry** [comment]
This will require another extremely annoying rebase in my VFS branch.

Keep Filefinder alone! 😭

**Comment by fmatthew5876** [comment]
> 
> 
> This will require another extremely annoying rebase in my VFS branch.
> 
> Keep Filefinder alone! 😭

I understand your pain, but this is an important fix.

**Comment by Ghabry** [comment]
Jenkins: test this please

**Comment by fdelapena** [comment]
Jenkins: test this please

**Comment by fdelapena** [comment]
I've tested the PR executable with #839, it does not fix that particular issue yet on Windows with the Blue Star Story game.

**Comment by fmatthew5876** [comment]
Can you post the 2 filenames in hex? The one player tries to open, and the actual filename on disk.

**Comment by carstene1ns** [comment]
Maybe we should use some lowercase function from ICU, I guess there is some dedicated for unistrings. Might be aware of such things.

**Comment by SmiVan** [comment]
@carstene1ns [UnicodeString::toLower()](http://www.icu-project.org/apiref/icu4c/classicu_1_1UnicodeString.html#ac7fb1eeca89ad489988ef9847dc5cf09) or [UnicodeString::toLower(Locale)](http://www.icu-project.org/apiref/icu4c/classicu_1_1UnicodeString.html#a69ee6c3a17a91b52bacd592bb4975c13) might be what we need then.

**Comment by Ghabry** [comment]
we already use icu::UnicodeString(str.c_str()).toLower() function in Normalize.
But you are Right, I forgot that lower and uppercasing depends on the language...

File: src/filefinder.cpp [hunk 1]
```diff
@@ -627,16 +628,16 @@ bool FileFinder::IsValidProject(DirectoryTree const & dir) {
 
 bool FileFinder::IsRPG2kProject(DirectoryTree const& dir) {
 	string_map::const_iterator const
-		ldb_it = dir.files.find(Utils::LowerCase(DATABASE_NAME)),
-		lmt_it = dir.files.find(Utils::LowerCase(TREEMAP_NAME));
+		ldb_it = dir.files.find(ReaderUtil::Normalize(DATABASE_NAME)),
+		lmt_it = dir.files.find(ReaderUtil::Normalize(TREEMAP_NAME));
```
Comment (Ghabry) [inline src/filefinder.cpp]: Here Normalize is unnecessary

File: src/filefinder.cpp [hunk 2]
```diff
@@ -627,16 +628,16 @@ bool FileFinder::IsValidProject(DirectoryTree const & dir) {
...
 
 	return(ldb_it != dir.files.end() && lmt_it != dir.files.end());
 }
 
 bool FileFinder::IsEasyRpgProject(DirectoryTree const& dir){
 	string_map::const_iterator const
-		ldb_it = dir.files.find(Utils::LowerCase(DATABASE_NAME_EASYRPG)),
-		lmt_it = dir.files.find(Utils::LowerCase(TREEMAP_NAME_EASYRPG));
+		ldb_it = dir.files.find(ReaderUtil::Normalize(DATABASE_NAME_EASYRPG)),
+		lmt_it = dir.files.find(ReaderUtil::Normalize(TREEMAP_NAME_EASYRPG));
```
Comment (Ghabry) [inline src/filefinder.cpp]: same here

File: src/filefinder.cpp [hunk 3]
```diff
@@ -854,7 +855,7 @@ bool FileFinder::IsMajorUpdatedTree() {
 			string_map mem = tree->sub_members["music"];
 			for (auto& i : mem) {
 				std::string file = mem[i.first];
-				if (Utils::EndsWith(Utils::LowerCase(file), ".mp3")) {
+				if (Utils::EndsWith(ReaderUtil::Normalize(file), ".mp3")) {
```
Comment (Ghabry) [inline src/filefinder.cpp]: and here

**Comment by Ghabry** [comment]
To figure out what the filefinder has in the file list you will need to dump it, best is printf Debugging.

good place is here: https://github.com/EasyRPG/Player/blob/81249d3413c704e607dda5a7cb167f97ffdcbf69/src/filefinder.cpp#L811

**Comment by carstene1ns** [comment]
@Ghabry: Yep, I just remembered reading what @SmiVan found about default locale and such.

**Comment by fdelapena** [comment]
After taking a look, comparing with some arbitrary locale looks ugly, because we should map some locale with the detected encoding.

What about [caseCompare](https://stackoverflow.com/a/38469016) instead?

**Comment by Ghabry** [comment]
This must be a Windows only problem, I just tested this PR against multiple stuff under Linux and it works as expected.

**Comment by Ghabry** [comment]
rebased + applied my review comments

**Comment by carstene1ns** [comment]
Let's break things.

**Pull request merged**